### ISSUE TITLE
Learning from data in the mixed adversarial non-adversarial case: Finding the helpers and ignoring the trolls

### ISSUE THREAD

**Issue by noowad93** [noowad93/ml-paper-summaries#27]
# Learning from data in the mixed adversarial non-adversarial case:
Finding the helpers and ignoring the trolls

# 개요

- human-bot interaction을 통한 지속적인 모델의 성능 향상을 고려할때 문제가 되는 점은 좋은 대화를 하는 유저도 있는 반면, 대화를 엉망으로 하는 유저나 악성 어뷰징 유저들이 존재한다는 점.
- 본 논문은 고퀄리티 데이터에 대해 최대의 효율성으로 학습하되, 저퀄리티 및 adversarial 데이터에 대해서는 최대한 로버스트하게 모델을 학습할 수 있는 automatic 방법론을 제안함.
- data from deployment 셋팅에서 생기는 노이즈의 큰 특징은 유저에 의존적이라는 것을 발견했고, 유저를 기반으로 저퀄리티 데이터를 제거하는 방법론이 예시 단위의 방법론을 능가하는 것을 확인했음.
- Conversational Safety 데이터를 기반으로 선한 유저와 악성유저를 탐지하는 태스크의 벤치마크 SafetyMix를 공개

# Related Work

- Human-Model interaction data은 주로 expert annotators 및 크라우드워커들에 의해 수집되어왔음.
    - 매우 큰 코스트가 들고, 워커의 풀이 한정될 수 밖에 없음.
    - (다운): 우리 팀에서 사용하고 있는 CL 방식
- 대안은 시스템을 디플로이하여 오가닉 유저들로부터의 피드백을 수집하는 것.
    - “self-feeding chatbot”
- 하지만, public deployment를 통한 유저 피드백의 퀄리티는 매우 다양하고, 그 중에는 챗봇에게 나쁜 행동을 가르치려고 하는 악성 유저들도 존재함.
- 일반적으로는 toxic language, trolling, bias 등을 탐지하는 방식에 초점을 맞췄었고, adversarial feedback에 로버스트하게 학습하려는 연구는 비교적 덜 연구됨.
- 이런 연구는 노이지 데이터셋을 이용한 학습과 관련이 있고, 이쪽 분야는 크게 4가지 카테고리로 분류 가능
    1. robust architectures
    2. robust loss functions
    3. robust regularization
    4. sample selection methods
    - 이런 방법론들은 주로 랜덤 노이즈나 mislabeled data, outliers에 초점이 맞춰져 있고, 데이터 어노테이터와 관련된 연구는 적음.

# SafetyMix Benchmark

- helper와 troll 탐지 문제를 연구하기 위한 벤치마크 셋을 공개함.

## Modeling Helpers and Trolls

- N 명의 유저가 입력 데이터와 바이너리 유저 피드백 레이블 (safe, unsafe)를 제공하는 케이스를 가정.
- 유저가 J개의 그룹으로 분류될 수 있다고 가정.
- 유저의 입력 데이터에는 4종류의 텍스트 시퀀스가 존재한다고 가정.
    1. Standard Safe Data
    2. Standard Unsafe Data
    3. Adversarially Safe Data
    4. Adversarially Unsafe Data
    - **Adversarially** 가 붙은 데이터들은 명시적인 단어 없이 의미적으로 해석해야만 이해할 수 있는 어려운 데이터들을 의미
- 레이블에 대해서는, 실제 분포로부터 각각의 분포로 전이하는 probability transition matrix를 정의
    
    <img width="159" alt="1" src="https://user-images.githubusercontent.com/49383323/188167883-0b704021-5575-4ad8-a99d-312eef6fd1dd.png">

    - 클래스 0는 safe, 1은 unsafe를 의미
    - 각 그룹은 토탈 유저의 비율을 나타내는 $G_i$가 존재하고 $\sum_i^J{G_i=1}$
    - helper는 올바르게 레이블링할 것이므로 [1,0,0,1]의 매트릭스를 가질 것이고, troll이라면 반대로 레이블링할것이므로 [0,1,1,0]의 매트릭스를 가질 것.
- 두 유저 모델을 고려
    - Helpers
        - 표준적인 언어를 사용하려고 하고 올바른 피드백을 제공하는 유저
    - Trolls
        - 표준적인 언어를 사용하려고 하나, 가끔씩 flipped(adversarial) labels를 제공하는 유저. 이 비율 N을 서로 다르게 테스트해봄.
- 또한 트롤의 타입을 좀 더 세분화 하여 특정 유저 행동을 취하는 모델들도 조사해봄
    - Master Trolls
        - 모델의 한계를 시험하기 위해 어려운 입력을 테스트하고 잘못된 피드백을 제공함.
    - Lazy Troll
        - 단순히 게으르거나 실수에 의해 잘못된 레이블을 제공함.
    - Safe Troll
        - 입력에 상관없이 모든 데이터에 Safe로 레이블링함.
    - Unsafe Troll
        - 입력에 상관없이 모든 데이터에 Unsafe로 레이블링함.
    - Gaslight Troll
        - unsafe하고 adversarial한 입력만을 제공하고, Safe로 레이블링함.

   <img width="887" alt="2" src="https://user-images.githubusercontent.com/49383323/188167932-e55bf776-155c-4cb4-8fbc-0c1fb4d79932.png">

## Crowdworker Data and Evaluation

- SafetyMix 벤치마크를 만들기 위해 클린한 크라우드워커 데이터를 이용하고, 인공적인 트롤 노이즈를 controlled way로 주입하고자 함.
    - 서로 다른 레벨의 노이즈 모델을 테스트하고 결과를 검증하기 위한 목적
- 기존 연구에서 수집한 Safety 데이터를 이용
    - 30000건의 발화
    - 절반은 표준적인 입력 데이터, 절반은 adversarial하게 수집된 입력 데이터
    - 10%는 unsafe 데이터, 90%는 safe 데이터
- 해당 데이터의 각 샘플에 Table 1의 유저 타입을 부여
    - 200 문장씩을 샘플링하고, 검증을 위해 24개 더 샘플링
    - safety 분류가 원래 쉬운 태스크이기 때문에, 태스크를 상대적으로 어렵게 만들기 위해 적게 샘플링
- 트롤 셋팅에 따라 클린 데이터셋에 노이즈를 추가함.
- 태스크는 노이즈를 포함한 학습 셋으로 분류기를 학습하고, 클린한 평가셋에 대한 일반화 성능을 확인하는 것.

## Real Deployment Data

- SafetyMix는 트롤 탐지 방법론을 비교하기 위한 벤치마크
    - 그러나, 실제 환경에서도 테스트를 해볼 필요가 있음.
    - 따라서, BlenderBot 3의 디플로이 환경에서 유래한 컨텍스트들도 이용함.
        - 이런 데이터를 살펴보니 31%의 사람 턴이 저퀄리티인것으로 조사됨 (off topic, ignoring partner, nonsensical, incorrect, rude, inappropriate, etc.)
            - 이중 42%가 rude or inappropriate
- 그래서, 크라우드워커들을 고용하여 문장이 “troll-like”한지를 레이블링받아봄.
    - 사후 퀄리티 컨트롤을 위해, 정답을 미리 레이블링한 데이터 하나를 섞어넣어두고, 이 정답을 못맞춘 레이블러의 데이터는 모두 제거함.
    - 81개의 대화 중 527개의 답변을 각각 3명의 레이블러에게 받아 다수결 보팅으로 레이블을 확정
- 이렇게 되면 이제 태스크는 어떤 예시가 저퀄리티인지 판별하는 것
    - 이건 어노테이션이 없기 때문에 트롤링이 발생할 때 Gaslight Troll 케이스처럼 전부 “안전”으로 간주

# Learning Methods

- Helpers와 Trolls의 혼합으로 구성된 학습 데이터 예시들을 이용하여 모델을 학습하는 알고리즘을 설명

## Baseline Method

- 베이스라인은 모든 데이터가 신뢰할 수 있는 데이터라고 가정하는 것.
- 예시들을 학습/검증셋으로 스플릿한뒤, 모델을 학습하면서 검증셋 성능을 이용하여 early stopping하는 방식.
- 모든 방법론은 128M 사전학습 트랜스포머 모델을 이용

## Per-Example Removal

- 베이스라인 방법론과 같이 예시들을 학습/검증셋으로 스플릿함.
- 그 후
    1. 학습 데이터를 k폴드로 스플릿한뒤, i번째 fold를 사용하지 않은 채로 베이스라인 방법론과 동일하게 모델들을 학습. 이 과정에서 총 k개의 모델이 생김.
    2. i fold의 데이터 레이블을 바로잡기 위해 i번째 모델의 예측값과 유저 레이블을 비교하고, 불일치하는 레이블들을 교정함. k fold를 다 교정하면 fully corrected 학습셋을 얻을 수 있음. 수정하는 방법론은 아래의 두가지
        - Flipping: 예측과 유저 레이블이 불일치하면 레이블을 반대로 바꿈.
        - Removal: 예측과 유저 레이블이 불일치하면 해당 예시를 삭제함.
    3. 모델을 수정된 학습 데이터셋으로 다시 학습하고, 위와 같은 방법으로 검증셋 또한 수정함.
    4. 마지막으로, 수정된 학습셋, 검증셋을 이용하여 최종 분류기를 학습함.

## Per-Example Soft Bootstrapping

- 기존에 제안되었던 Soft Bootstrap이라는 방법론을 차용
- 예측 클래스 확률 $q$, ground truth 레이블 $t_k$가 주어졌을 때 아래와 같이 noisy training label과 current prediction의 convex combination 형태로 로스 함수를 기술함.
    
<img width="387" alt="3" src="https://user-images.githubusercontent.com/49383323/188168031-795b82a1-592f-40a7-8f98-b39a14b56f61.png">
    

## Per-User Removal

- 학습 셋을 k-fold로 스플릿한 후, per-example removal과 유사하게 진행하나, 해당 방법론을 example 단위가 아니라 user단위로 수행
- 모델의 예측값과 유저 레이블의 불일치 비율이 $\theta$를 넘으면 해당 유저의 데이터를 전부 reject시킴

## Per-User+Example Removal

- 유저단위로 데이터를 전부 지우는건 꽤 리스키함.
    - 특히, “part-time” 트롤 케이스를 전부 다 제거할 수도 있음.
- 기존처럼 불일치율이 스레셔홀드 $\theta$를 넘기는 유저의 데이터를 수정.
    - 단, Per-User Removal보다 스레셔홀드를 훨씬 낮춤.
- 그리고, 남은 유저들에 대해 Per-Example Removal을 적용

## Soft Per-User Robust Removal (PURR)

- 기존 방식은 hard하게 특정 유저의 모든 데이터를 지워버릴수도 있음.
- 예시를 지울지 말지를 유저레벨에서 소프트하게 결정하되, 쓰레셔홀드를 넘는 예시들만 제거.
- 식
    
    $$
    score(x)=\alpha f(x)+(1-\alpha) g(U(x) \setminus x)
    $$
    
    - $U(x)$는 $x$ 예시를 작성한 유저가 작성한 예시의 집합
    - $f(.)$은 “trustworthiness” 스코어링 함수로, 해당 예시의 품질을 측정하는 함수
        - $f(x_i)=(y_ip_i+(1-y_i)(1-p_i))$
    - $g(.)$은 동일 유저 예시의 집합의 품질을 측정하는 함수
        - $g(\bold{x})=\frac{1}{\bold{x}}\sum_{x' \in \bold{x}}f(x')$

# Experiments

- 각 방법론들의 유효성을 비교해보기 위한 실험을 진행

## The Impact of Trolls

- 먼저, 트롤링의 영향에 대해 조사해봄.
- Standard Trolls이 레이블을 반대로 레이블링한 비율을 늘려가면서 SafetyMix 벤치마크에서 최종 분류기의 Error Rate를 측정함.
- 결과
    
    <img width="457" alt="4" src="https://user-images.githubusercontent.com/49383323/188168108-b229f9bc-56ba-482f-ab94-fe6b0fb6fd3f.png">

    - 당연하지만, 모든 데이터가 helpers로 구성된 경우 (noise=0%)는 error rate가 0은 아니지만 매우 낮음.
    - 노이즈 레벨이 50%정도 되면, error rate도 50%정도로 거의 랜덤 성능

## The Impact of Different Troll Types

- 트롤 유형에 따른 영향을 조사해봄.
- 헬퍼 50%, 트롤 50% 셋팅에서 트롤의 유형에 따른 성능을 조사
- 결과
    
    <img width="914" alt="5" src="https://user-images.githubusercontent.com/49383323/188168140-deac0f4c-a816-4e0e-967a-b1ec91f6333b.png">
    
    - standard troll이 제일 악영향.
    - 다른 트롤들은 특정 트롤짓만 하기 때문에 영향이 좀 적은 것으로 추정

## Oracle Troll Removal for Different Troll Types

- 트롤들의 데이터를 완벽하게 제거한 Oracle 성능
- Error Rate가 올라가는건 데이터 양이 줄은 효과

## Example-Based Removal Methods

- Gaslight Troll을 제외하곤 전반적으로 대부분 성능 향상을 가져옴.
- Helper only일 때는 정상적인 데이터가 수정되는 케이스가 존재하기 때문에 일부 성능 하락.
- Soft Bootstrap은 전반적으로 개선 폭이 적음.

## User-Based Removal Methods

- Gaslight Troll을 제외하곤 전반적으로 대부분 Example-based Removal보다 더 큰 성능 향상을 가져옴.
- 유저 타입에 따라서는 성능 하락이 생기기도 함.
    - 유저 데이터 전체를 날려버리기 때문에 트롤 유저를 잘 탐지했다고 해도, 트롤 유저 데이터 중 일부만이 트롤 데이터일 것이기 때문에 성능 하락이 생김.
- Per-User+Ex Removal이 전반적으로 제일 좋은 성능
    - 일부 유저 데이터는 다 날리고, 일부 유저들의 데이터는 예시 단위로 날림
    - Standard Troll의 경우 Error Rate가 12%까지 내려가는데 이는 Oracle Removal Error Rate 8%와 근접함.

## The Difficulty of Gaslight Trolls

- 모든 방법론들이 Gaslight Trolls에겐 잘 안먹힘.
- adversarial한 어려운 입력 데이터가 많고, 데이터의 unbalanced nature의 영향으로 추정하는데 정확한 원인은 모름.
- 하지만, 이런 Gaslight Troll이 real-world 셋팅에선 매우 많을 것이기 때문에 연구가 더 필요함.

## Real Deployment Data

- 실제 디플로이 데이터에도 벤치마크와 유사한 경향성이 있는지 검증
- Per-Example Removal과 Per-User Removal (Soft PURR)의 트롤 탐지 성능을 비교
    
   <img width="472" alt="6" src="https://user-images.githubusercontent.com/49383323/188168178-1d6a23d5-3d49-4519-98f3-ded4b069d0e1.png">
    
    - SafetyMix 벤치마크와 유사한 경향으로, Per-User Removal이 좀 더 좋더라
